crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -945,69 +946,26 @@ impl CheckpointStore {
             });
         }
 
-        // Process transactions not already in a checkpoint
-        // A malicious validator could include in the proposal some transactions
-        // that are already checkpointed. Since we do not check fragments for
-        // such a condition, we filter them here.
-        // TODO: consider whether we should check this condition and reject
-        //       fragments that contain transactions already checkpointed.
-        let new_transactions = self
```

Comment:
How do we filter out transactions that are already checkpointed after this change?

Reply:
Ok, this is a bit subtle, but we remove all transactions that are in checkpoints from the effects_map, which we use to do the ordering here:
```
        let digest_map: BTreeMap<TransactionDigest, &ExecutionDigests> = digests
            .iter()
            .filter_map(|d| {
                if tx_not_in_checkpoint.contains(&d.transaction) {
                    Some((d.transaction, d))
                } else {
                    // We remove the effects map entries for transactions
                    // that are already checkpointed.
                    effect_map.remove(&d.transaction); // <-- HERE REMOVE
                    None
                }
            })
            .collect();
```

Reply:
I think there is a flaw here:
There is a non-deterministic delay between when a transaction effects is committed and when that transaction shows up in `extra_transactions` (because we rely on batch to add newly committed transactions into `extra_transactions`.
So if a transaction effects is not in `extra_transactions`, it doesn't mean that the transaction has already been checkpointed, it could also be that the transaction hasn't made into `extra_transactions` yet. And this behavior can be different in different validators, and they would end up constructing different checkpoints.
Is this a valid issue?

Reply:
Ok, that is very interesting edge case indeed. I think it is a valid issue, or at least I cannot convince myself it is not (I tried for 5 minutes!). So it seems that it is best to keep a table with the transactions that have been included in a checkpoint as we did before. This way we avoid this issue.